### PR TITLE
add/support enabled_tenants and disabled_tenants feature in storegateway

### LINKED ISSUE #5133: Support enabled_tenants and disabled_tenants in store-gateway
**Is your feature request related to a problem? Please describe.**
enabled_tenants and disabled_tenants in rulers, compactors and alertmanager is a nice feature to run different ruler and compactors and alertmanager for different tenants.
It should be possible to do the same in store-gateway

**Describe the solution you'd like**
It should be possible to define -store-gateway.enabled-tenants and -store-gateway.disabled-tenants to enable specific tenants in the store-gateway cluster

**Describe alternatives you've considered**
None, I think this one is consistent with the rest of the options.

**Additional context**
None

### PR DESCRIPTION
**Pull request by moki1202**
Signed-off-by: Shashank [shashanksingh819@gmail.com](mailto:shashanksingh819@gmail.com)

What this PR does: Now adds/supports `enabled_tenants` and` disabled_tenants` in package `storegateway`

Fixes #5133

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

PR still in progress.

### CODE REVIEW

**Comment by moki1202** [comment]
@friedrich-at-adobe Do I need to create a new function to filter out users that are not allowed?

**Comment by friedrichg** [comment]
@moki1202 Store-gateways download the bucket index for the tenant and all the blocks. I would expect that this PR to ignore users discovered in S3 that are not allowed. a new function is not required, use `allowedTenants.IsAllowed`
See this [PR](https://github.com/cortexproject/cortex/pull/5116) for an example

**Comment by friedrichg** [comment]
@moki1202 filterUsers function is already in the right places
https://github.com/cortexproject/cortex/blob/47733ddab802acef66f6ab7be41ebb24a512a7ce/pkg/storegateway/bucket_index_metadata_fetcher.go#L61

https://github.com/cortexproject/cortex/blob/47733ddab802acef66f6ab7be41ebb24a512a7ce/pkg/storegateway/bucket_stores.go#L212

Concentrate making changes on https://github.com/cortexproject/cortex/blob/master/pkg/storegateway/sharding_strategy.go
to implement this feature on both normal sharding and shuffle sharding
and don't forget to update/include tests in sharding_strategy_test.go

**Comment by moki1202** [comment]
@friedrichg Apologies for the late follow up. I've written a test for the changes. Can you please confirm whether its the correct way or not? Once I get the confirmation I'll try completing this ASAP.

**Comment by moki1202** [comment]
@friedrichg I've tried a lot but I am struggling while updating the tests. Can you please help me out here? I would just ask for your help in updating the defaultShardingStrategy tests. I'll try doing the other ones after.

**Comment by moki1202** [comment]
@friedrichg I've tried a lot but I am struggling while updating the tests. Can you please help me out here? I would just ask for your help in updating the defaultShardingStrategy tests. I'll try doing the other ones after.

**Review by friedrichg — CHANGES_REQUESTED** [review]

File: pkg/storegateway/sharding_strategy_test.go [hunk 1]
```diff
@@ -64,6 +67,21 @@ func TestDefaultShardingStrategy(t *testing.T) {
 				"127.0.0.2": {block2, block4},
 			},
 		},
+		"two ACTIVE instances in the ring with replication factor = 1 and one tenant disabled": {
```
Comment (friedrichg) [inline pkg/storegateway/sharding_strategy_test.go:70]: Unfortunately, this function does not test FilterUsers, it tests FilterBlocks. So this is not the right place

**Comment by friedrichg** [comment]
Thanks, partially included in https://github.com/cortexproject/cortex/pull/5638

**Pull request closed without merging**